[Draft] Second version of cupy support for sparse matrix cg sampling

chinandrew opened the pull request:
Unlike #15, this PR adds the use_cupy flag to the model/matrix, and then has the dot function handle the types of the input vectors case by case. This way cg_sampler can just use the design matrix flags without needing a use_cupy flag. Theres' probably cleaner way to do the logic, but as a prototype it seems to work and is actually faster than #15.

aki-nishimura commented:
I refactored the code a bit, but otherwise the cupy integration looks good to me. Perhaps you can go ahead and ff-merge after squashing commits to clean things up? It seems reasonable to just squash all the commits since `master` into one unless you see smaller coherent units.

Incidentally, an idea for even cleaner approach going forward. It seems like we have managed to completely get rid of the scipy matrix when using cupy? In this case, instead of having the `use_cupy` flag, maybe the `RegressionModel()` should just accept cupy matrix as an input and the whole module should adjust its behavior based on the type of `X`.  The `self.use_cupy` could even be a property that checks if `X` is a cupy array or not.

chinandrew commented:
> I refactored the code a bit, but otherwise the cupy integration looks good to me. Perhaps you can go ahead and ff-merge after squashing commits to clean things up? It seems reasonable to just squash all the commits since `master` into one unless you see smaller coherent units.


Yea squashing works for me when it's ready.

> Incidentally, an idea for even cleaner approach going forward. It seems like we have managed to completely get rid of the scipy matrix when using cupy? In this case, instead of having the `use_cupy` flag, maybe the `RegressionModel()` should just accept cupy matrix as an input and the whole module should adjust its behavior based on the type of `X`. The `self.use_cupy` could even be a property that checks if `X` is a cupy array or not.


I just tried giving this a shot by adding a `is_cupy_sparse` method to abstract matrix (wasn't quite sure where the best place to put it was, open to suggestions) that gets called by the sparse design matrix and also by the factory. Also updated some various logic to handle cupy input matrices (e.g. removing the intercept). Let me know what you think, we can always revert back to the previous commit if we think it needs more work.

chinandrew commented on bayesbridge/design_matrix/abstract_matrix.py:
```diff
@@ -63,12 +70,16 @@ def toarray(self):
-            col_variance = np.squeeze(np.array(
-                X.power(2).mean(axis=0) - np.power(X.mean(axis=0), 2)
-            ))
+        squeeze, array, power = (cp.squeeze, cp.array, cp.power) if \
```
at some point it might be easier to doing something like `linalg_pkg = cp if self.use_cupy else np`...

aki-nishimura replied:
`linalg_pkg = cp if self.use_cupy else np` is a great idea. (The name `linalg_pkg` might be confusing, though; it sounds like a type of BLAS. Off top of my head, `array_lib` comes to my mind, though still not great.)

chinandrew commented:
Did some minor updates to logics/checks around use_cupy to catch edge cases and clean up messages, added some basic gpu-dependent tests (pass on both my machine and jhpce),  and added cupy as an extra optional dependency to setup.py

chinandrew commented on bayesbridge/design_matrix/sparse_matrix.py:
```diff
@@ -73,7 +68,7 @@ def dot(self, v):
             self.v_prev = v.copy()
 
-        input_is_cupy = (cp is not None) and isinstance(v, cp._core.core.ndarray)
+        input_is_cupy = (cp is not None) and isinstance(v, cp.ndarray)
```
turns out this works

aki-nishimura commented on bayesbridge/design_matrix/abstract_matrix.py:
```diff
@@ -63,12 +70,16 @@ def toarray(self):
+        squeeze, array, power = (cp.squeeze, cp.array, cp.power) if \
+            AbstractDesignMatrix.is_cupy_sparse(X) else (np.squeeze, np.array, np.power)
+        if sp.sparse.issparse(X) or AbstractDesignMatrix.is_cupy_sparse(X):
+            col_variance = squeeze(array(X.power(2).mean(axis=0) - power(X.mean(axis=0), 2)))
```
Would the following work (w/o `{np, cp}.power()`)?
```suggestion
            col_variance = squeeze(array(X.power(2).mean(axis=0) - X.mean(axis=0).power(2)))
```

chinandrew replied:
I don't think so, `power()` doesnt appear to be a method for matrices (for np and cp).
```
In [10]: X = np.ones((5,5))

In [11]: X.mean(axis=0).power(2)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-11-2fff9e220fa4> in <module>
----> 1 X.mean(axis=0).power(2)

AttributeError: 'numpy.ndarray' object has no attribute 'power'
```

aki-nishimura replied:
Ah, silly me. Thanks, good to know.

chinandrew commented on tests/gpu_tests/test_gibbs.py:
```diff
@@ -1,12 +1,18 @@
 from ..helper import simulate_data
 
 
+pytestmark = pytest.mark.skipif(cp is None, reason="cupy unavailable")
```
Good call to use pytest skip functionality. Probably easiest to just put `pytest.importorskip("cupy")` at the top of the file, just pushed a commit for it.
